qringbuf.go

```diff
@@ -1,3 +1,168 @@
+//
+// In code the basic usage looks roughly like this:
+//
+//  qrb, initErr := qringbuf.NewFromReader( *someReader, qringbuf.Config{ … } )
+//  …
+//  readLimit = int64( stopReadingAfterThatManyBytes )
+//  startErr := qrb.StartFill( readLimit )
+//  …
+//  var available, processed int
+//  for {
+//      reg, streamErr := qrb.NextRegion( available - processed )
+//      if reg == nil {
+//          return streamErr
+//      }
+//
+//      // work with region, processing all or just a portion of the data
+//      available = reg.Size()
+//      processed = frobnicate( reg.Bytes(), … )
+//  }
+//
+// In addition one can operate over individual (sub)regions with "fearless
+// concurrency":
+//
+//  …
+//  var available, processed int
+//  for {
+//      reg, streamErr := qrb.NextRegion( available - processed )
+//      if reg == nil {
+//          return streamErr
+//      }
+//
+//      available = reg.Size()
+//      processed = 256
+//
+//      subReg := reg.SubRegion( 0, 256 )
+//      subReg.Reserve()
+//
+//      go func() {
+//          frobnicate( subReg.Bytes(), … )
+//          subReg.Release()
+//      }()
+//  }
+//
+// In all cases the background collector goroutine reading from *someReader
+// into the ring buffer is guaranteed to:
+//  - never overwrite the buffer portion backing the latest result of NextRegion()
+//  - never overwrite any buffer portion backing a Reserve()d (Sub)Region
+//
+// Implementation notes
+//
+// Unlike io.ReadAtLeast, errors from the underlying reader are always made
+// available on NextRegion(). As with the standard io.* semantics an error can
+// be returned together with a result. One should always check whether the
+// *Region return value is nil first, before processing the error,
+//
+// Changes of the emitter and collector positions are protected by a mutex on
+// the qringbuf object. Any call accessing the buffer state will block until
+// this lock can be obtained. One can also lock the mutex if the collector
+// goroutine needs to be paused for whatever reason.
+//
+// The region Reserve()/Release() functionality does not use the mutex, ensuring
+// that an asynchronous Release() call can not be affected by the current state
+// of the buffer. Reservation tracking is implemented as an atomically modified
+// list of reservation counts, one int32 per SectorSize bytes of the buffer.
+//
+// The reservation system explicitly allows "recursive locking": you can hold
+// an arbitrary number of reservations over a sector by repeatedly creating
+// SubRegion() objects. Care must be taken to release every single reservation
+// obtained previously, otherwise the collector will remain blocked forever.
+//
+// Here is an illustration of a qringbuf object lifecycle initialized with
+//  {
+//    BufferSize: 64,
+//    MinRegion:  16,
+//    MinRead:    8,
+//    MaxCopy     24,
+//  }
+// ·
+//
+//  † C is the collector position: the *end* of the most recent read from the underlying io.Reader
```

Comment:
Cool diagrams 👍 

I think this section may be a little clearer if it had an example showing the various function calls, and then the diagrams refer to those line numbers.

Reply:
Thanks! Took way longer than I initially thought it will, but actually uncovered an inconsistency while I was drawing them, so all is well.

I am not entirely sure what you are suggesting. Perhaps discussing this voice some time today would help?

Reply:
What I mean is that the numbered items seem to be describing a piece of code:
⓪ Buffer initialized, StartFill(…) is called ...
① NextRegion(0) is blocked until ...

So it might be clearer if there's a concrete example of that code to compare with:
Line 3: Buffer initialized, StartFill(…) is called ...
Line 5: NextRegion(0) is blocked until ...

Happy to talk on the phone if that's not clear :)

Reply:
Ohhhhh I see. This... wouldn't be... something that can be shown in docs, as the diagrams show the "tricky bits" which generally arise quite seldomly if ever. In fact the only way I managed to test all this is by essentially fuzzing the library by spinning it on an overloaded machine starved of CPU time.

I added a bit more prose in https://github.com/ipfs/go-qringbuf/commit/cb4888ddf8db, below is what it looks like visually. If I misunderstood what you are suggesting: let's chat after the launch meet
<img width="656" alt="image" src="https://user-images.githubusercontent.com/356110/80816775-71631a00-8bd0-11ea-9af0-fd9477d05881.png">

Reply:
Ok no worries, I don't think you need the disclaimer - it was just a suggestion to add some context to the diagrams, but I think it's fine as is 👍

Reply:
No, the disclaimer is good to have - if anything it puts things in perspective. So I'll def keep it
Thanks for all your suggestions, things are *WAY* better now ;)